### PR TITLE
feat(routine): add dfs and bfs descendant search methods with tests

### LINKED ISSUE #185: Add convenience functions for finding subroutines
## Is your feature request related to a problem? If so, please describe it.
Finding specific subroutines requires in-depth knowledge of the routine itself, and which routine has which child and so on. While the author will have some indication of what lives where, the precise pathway to a specific routine does not always come to mind easily. 

## How you envision the solution
Convenience functions written in the [`BaseRoutine` class](https://github.com/PsiQ/bartiq/blob/03ccaec64e598bc245d90f7efa241516cc237196/src/bartiq/_routine.py#L94) that enable quick traversal through the routine graph. For example, `.find_descendant(name)` could be a method that returns pathways to all descendants that have the name `name` (as subroutine names are non-unique). 

i.e. for a routine like:
```
root
├─ A
│  ├─ X
│  ├─ Y
├─ B
│  ├─ X
│  ├─ Y
│  ├─ Z
```
`find_descendant(X)` would return:
```
(['A', 'X'], ['B', 'X'])
```

Similarly:
```
find_descendant('Z')
>>> (['B', 'Z'])
```
where each entry in the returned tuple would be a pathway to the routine with that name, i.e.
```
Z: Routine = root.children('B').children('Z') 
```

`get_descendant` would be a similar method that takes in either a name or pathway returned from `find_descendant`, and returns the actual `Routine`/`CompiledRoutine` object. 

## Describe alternatives you've considered
I wrote some basic code that does the above functionality but only for uniquely named subroutines, so it does not work for generic routines. These were written as methods of `BaseRoutine`; they rely only on subroutine names and so will work for either `Routine` or `CompiledRoutine`. 

Moreover, this code only performs depth first search, and perhaps a kwarg should be to alternate between DFS and BFS. 

```
    def find_descendant(self, name: str) -> list[str]:
        """Find the pathway to a descendant subroutine given its name.

        At present this is done via depth-first search only.

        Args:
            name: The name of the desired subroutine.

        Returns:
            list[str]: A list of children names where the final entry is the input name, representing
                    graph edges to traverse.
        """

        def recursive_search(routine: BaseRoutine, pathway: Iterable[str] | None = None) -> tuple[list[str], bool]:
            pathway = pathway or []
            found = False
            for child_name, child_routine in routine.children.items():
                if "qubits" in child_name:
                    continue
                if child_name == name:
                    pathway.append(child_name)
                    return pathway, True
                else:
                    pathway, found = recursive_search(child_routine, pathway)
                    if found:
                        pathway.append(child_name)
                        return pathway, found
            return pathway, found

        pathway, found = recursive_search(self)
        if not found:
            raise ValueError(f"No routine {name} found in the descendants of {self.name}.")
        return pathway[::-1]

    def get_descendant(self, name: str) -> BaseRoutine:
        """Get the descendant of the parent routine given its name.

        Does not have to be a direct child.

        Args:
            name: Name of the desired routine.

        Returns:
            BaseRoutine
        """
        pathway = self.find_descendant(name=name)
        subroutine = self
        for subroutine_name in pathway:
            subroutine = subroutine.children[subroutine_name]

        return subroutine
```

## Additional context
Perhaps conversion from routine to a graph object ( a la networkx) would streamline this process, as traversal across python graphs is well-trodden ground. There are potentially more use cases for a converting a routine into a graph object, too

### PR DESCRIPTION
**Pull request by poig**
## Description

trying to close #185 

Navigating and querying deeply nested quantum routines is cumbersome, especially when subroutine names are non-unique or the structure is complex. Previously, there was no convenient way to find all descendants with a given name, limit search depth, or retrieve the closest match.

Implemented Solution:

- Added a flexible find_descendants method to BaseRoutine that supports:

  - Both DFS recursive and BFS iterative traversal modes.
  - Early exit with first_only for efficient nearest-match queries.
  - Depth limiting via max_depth.
  - Option to return both the path and the routine object (return_objects).

- Added/expanded comprehensive tests in test_routine.py to cover all traversal modes, options, and edge cases.
- The new utilities make it easy to find, filter, and retrieve subroutines by name and path, improving developer productivity and code clarity.

should we make a tutorial introducing these feature?

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

File: src/bartiq/_routine.py [hunk 1]
```diff
@@ -145,6 +145,78 @@ def sorted_children_order(self) -> Iterable[str]:
     def sorted_children(self) -> Iterable[Self]:
         return [self.children[child_name] for child_name in self.sorted_children_order]
 
+    def find_descendants(
+        self,
+        name: str,
+        mode: str = "dfs",
```
[the remark this answers is not in the document]
Comment (poig) [reply src/bartiq/_routine.py]: thanks for suggestion, I modified it

**Comment by poig** [comment]
> Really nice! Just a few comments, let me know if you need clarification on anything

just fixed some syntax issue, I think its ready to merge

**Comment by mstechly** [comment]
@poig Could you please send me an e-mail at mstechly@psiquantum.com, so that I can send you a CLA to sign?

**Comment by poig** [comment]
> @poig Could you please send me an e-mail at [mstechly@psiquantum.com](mailto:mstechly@psiquantum.com), so that I can send you a CLA to sign?

sent a email

**Comment by poig** [comment]
> Looks good, once all the comments are resolved I think this is ready to merge.

I think resolved all the comments, have a look again please, thank you!

**Review by Brendan-Reid1991 — APPROVED** [review]
Sorry @poig ! I think one of my comments wasn't super clear, but just make the two `_depth_first_search` and `_breadth_first_search` functions standalone rather than inner, and we can merge

File: src/bartiq/_routine.py [hunk 2]
```diff
@@ -145,6 +145,64 @@ def sorted_children_order(self) -> Iterable[str]:
...
+        Args:
+            name: The name of the desired subroutine.
+            mode: 'dfs' (default) or 'bfs' for traversal order.
+            max_depth: Maximum depth to search (None = unlimited).
+
+        Returns:
+            List of pathways (each a list of child names).
+        """
+
+        def _depth_first_search(routine: Self, path: list[str], depth: int) -> list[list[str]]:
```
[the remark this answers is not in the document]
Comment (poig) [reply src/bartiq/_routine.py]: I just modified both function to the bottom of the file, thank you for the review
